### DISCUSSION TITLE
type annotation which is ignored by pyright

### DISCUSSION THREAD

**Discussion by nonagonal** [microsoft/pyright#7459] [General]
I'm adding type annotations to global variables and have special code to process them as metadata at runtime. I'd like to have these annotations ignored for the purposes of static type checking.

So in a utility module I have:
```
type MySpecialType = Any
```

and then in dozens of other modules I have:
```
from myspecialmodule import MySpecialType
...
EXAMPLE_VARIABLE: MySpecialType = example_value

```

For pyright's purposes I don't actually want to annotate `EXAMPLE_VARIABLE` as `Any`. Is there a type I can use e.g. if `typing.TYPE_CHECKING` is `True` in place of `Any` to get pyright to effectively ignore annotations that use it?

**Comment by erictraut** [comment]
You're using annotations in an unconventional and non-conformant manner, so your options will be rather limited here. Here are a couple of ideas.

1. You could specify the _actual_ type of the constant (e.g. `int` or `str`) in conjunction non-typing metadata using `typing.Annotated`. Your runtime logic an then pick up the non-typing metadata from the `Annotated` object. For more details on `Annotated`, see [this documentation](https://typing.readthedocs.io/en/latest/spec/qualifiers.html#annotated). 

2. You could define `MySpecialType` as an alias of the `Final` qualifier when `TYPE_CHECKING` is true.
```python
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from typing import Final as MySpecialType
else:
    MySpecialType = int

TEST: MySpecialType = 4
```

Pyright's ability to alias `Final` is relatively limited, so this technique would require you to place the above code in every file that uses `MySpecialType` rather than placing it in a common shared file.

**Reply by nonagonal** [reply]
I was hoping for a special non-type type, but it sounds like the Python community has moved on from non-type applications of type annotations.

It's unfortunate because I think with extremely minimal tweaks (e.g. allow a form of Annotated which simply attaches metadata and says nothing about type) it opens up many exciting applications! I'll try to stick with other mechanisms, I just can't find a syntax for attaching metadata to variables that's anywhere near as elegant as type annotations.

**Comment by Azureblade3808** [comment]
Try this --

```python
from typing import Protocol

class MySpecialType(Protocol):
    pass

TEST: MySpecialType = 4
reveal_type(TEST)  # "Literal[4]"
```

--

Edit:
Type of `TEST` will be regarded as `MySpecialType` instead of `Literal[4]` in another module, so this snippet probably doesn't meet your need.

**Reply by nonagonal** [reply]
Oh neat - my use of TEST is limited to the module in which it's defined, so this would indeed work for me. Why is it that type checkers infer a different type when TEST is imported from another module?

**Reply by Azureblade3808** [reply]
See https://github.com/microsoft/pyright/blob/main/docs/type-concepts-advanced.md#type-narrowing for some accurate info.

Below is my own understanding:

As for variable `TEST`, its declared type is `MySpecialType`, and its actual type is narrowed as `Literal[4]` after the assignment.

In the same module/file, Pyright will see that the value of `TEST` never gets modified after the first assignment, so its actual type can remain `Literal[4]` even in non-global scopes.

In other modules/files, assignment(s) to `TEST` is considered local implementation and is ignored (so that impact from local changes could be minimal), so its declared type gets used.
